src/Identity/Extensions.Core/src/UserManager.cs

```diff
@@ -2128,6 +2143,92 @@ public virtual Task<int> CountRecoveryCodesAsync(TUser user)
+    {
+        ThrowIfDisposed();
+        var passkeyStore = GetUserPasskeyStore();
+        ArgumentNullThrowHelper.ThrowIfNull(user);
+        ArgumentNullThrowHelper.ThrowIfNull(passkey);
+
+        await passkeyStore.SetPasskeyAsync(user, passkey, CancellationToken).ConfigureAwait(false);
+        return await UpdateUserAsync(user).ConfigureAwait(false);
+    }
+
+    /// <summary>
+    /// Gets a user's passkeys.
+    /// </summary>
+    /// <param name="user">The user whose passkeys should be retrieved.</param>
+    /// <returns>
+    /// The <see cref="Task"/> that represents the asynchronous operation, containing a list of the user's passkeys.
+    /// </returns>
+    public virtual Task<IList<UserPasskeyInfo>> GetPasskeysAsync(TUser user)
+    {
+        ThrowIfDisposed();
+        var passkeyStore = GetUserPasskeyStore();
+        ArgumentNullThrowHelper.ThrowIfNull(user);
+
+        return passkeyStore.GetPasskeysAsync(user, CancellationToken);
+    }
+
+    /// <summary>
+    /// Finds a user's passkey by its credential id.
+    /// </summary>
+    /// <param name="user">The user whose passkey should be retrieved.</param>
+    /// <param name="credentialId">The credential ID to search for.</param>
+    /// <returns>
+    /// The <see cref="Task"/> that represents the asynchronous operation, containing the passkey if found; otherwise <see langword="null"/>.
+    /// </returns>
+    public virtual Task<UserPasskeyInfo?> GetPasskeyAsync(TUser user, byte[] credentialId)
+    {
+        ThrowIfDisposed();
+        var passkeyStore = GetUserPasskeyStore();
+        ArgumentNullThrowHelper.ThrowIfNull(credentialId);
+
+        return passkeyStore.FindPasskeyAsync(user, credentialId, CancellationToken);
+    }
+
+    /// <summary>
+    /// Finds the user associated with a passkey.
+    /// </summary>
+    /// <param name="credentialId">The credential ID to search for.</param>
+    /// <returns>
+    /// The <see cref="Task"/> that represents the asynchronous operation, containing the user if found, otherwise <see langword="null"/>.
+    /// </returns>
+    public virtual Task<TUser?> FindByPasskeyIdAsync(byte[] credentialId)
+    {
+        ThrowIfDisposed();
+        var passkeyStore = GetUserPasskeyStore();
+        ArgumentNullThrowHelper.ThrowIfNull(credentialId);
+
+        return passkeyStore.FindByPasskeyIdAsync(credentialId, CancellationToken);
+    }
+
+    /// <summary>
+    /// Removes a passkey credential from a user.
+    /// </summary>
+    /// <param name="user">The user whose passkey should be removed.</param>
+    /// <param name="credentialId">The credential id of the passkey to remove.</param>
+    /// <returns>
+    /// The <see cref="Task"/> that represents the asynchronous operation, containing the <see cref="IdentityResult"/>
+    /// of the operation.
+    /// </returns>
+    public virtual async Task<IdentityResult> RemovePasskeyAsync(TUser user, byte[] credentialId)
```

Comment:
For all these methods, maybe use `ROM<byte>` unless we have a good reason to pass byte[] around?

Reply:
Ah, it's not possible because expression trees can't reference ref structs, and you need to access `ReadOnlyMemory<byte>.Span` when doing the ID equality check in the query for the `DbSet`.

We could still change the `UserManager` APIs to work with `ReadOnlyMemory<byte>`, but we'd need to eventually copy the buffer to a `byte[]` anyway. Keeping the argument as a `byte[]` enables the caller to avoid this unnecessary copy, especially if they already have a `byte[]` to provide in the first place.